_posts/2023-04-10-opensource-perf.markdown

```diff
@@ -0,0 +1,128 @@
+---
+layout: post
+title:  "Collaborating to Improve Open Source Performance: A Case Study"
+authors:
+  - dblock
+  - chibenwa
+date:   2023-04-10
+categories:
+  - technical-post
+meta_keywords: OpenSearch, Apache James, Flame graph, Performance, Driver
+meta_description: Learn how key contributors of APache James found out and fixed major performance problems in OpenSearch Java driver.
+---
+
+As an open source enthusiast, I believe in the power of collaboration to make open source projects faster and more efficient. In this blog post, I will share how my team at [LINAGORA](https://linagora.com), contributing to the Apache James project, identified and addressed a performance issue in OpenSearch using benchmark tools and flamegraphs, and collaborated with the OpenSearch community to make the project better.
+
+[Apache James](https://james.apache.org), a.k.a. Java Apache Mail Enterprise Server is an open source email server written in Java. It implements commons email protocols like SMTP, IMAP or [JMAP](https://jmap.io). Apache James is easy to customize with a wide range of extension mechanisms. It is even easy to use James as a toolbox to assemble your own email server! Apache James also proposes a novative architecture for email server: James is stateless and relies on NoSQL databases and message brokers for state management. Thus administrating James is as easy as managing any modern web applications: no sharding, no rotocol-aware load balancing. Apache James uses OpenSearch for search related features in the distributed mail server setup. 
+
+Operating an email server, performance is a concern, as emails are massively used applications, generaly considered critical. Often with even medium sized deployment, handling over 1000 requests per seconds is not uncommon. We use a range of tools for benchmarking the performance of Apache James. Our docs [explains](https://github.com/apache/james-project/blob/master/server/apps/distributed-app/docs/modules/ROOT/pages/benchmark/index.adoc) some of the benchmarks we suggest our administrators run to identify performance bottlenecks and malfunctions in their clusters.
+
+Apache James relies on OpenSearch as its search engine for its email database. It takes as a dependency the [opensearch-java client](https://github.com/opensearch-project/opensearch-java). Performance testing revealed that the application was spending a noticeable chunk of time running the opensearch-java code, so we set out to examine the performance characteristics of this component.
```

Comment:
Could you maybe clarify how this was detected? As in, which particular benchmark or analysis revealed this disparity?